### ISSUE TITLE
Bug in selective scan backward

### ISSUE THREAD

**Issue by Hprairie** [state-spaces/mamba#368]
Hey, I think I found a bug in selective_scan_bwd_kernel.cuh . When we are propagating gradients backward we are moving in a reverse scan and thus it's important of $\exp(\Delta A)$ to be 1 for threads with items over the sequence length. This will be the case when `delta_softplus=False` and `delta_bias=None`, however, this will not be the case when either of these is enabled. We can see that this true by looking at the following lines of code.

```CUDA
        load_input<Ktraits>(delta, delta_vals_load, smem_load, params.seqlen - chunk * kChunkSize);
        // Will reload delta at the same location if kDeltaSoftplus
        if constexpr (!kDeltaSoftplus) { delta -= kChunkSize; }

        float dout_vals[kNItems], delta_vals[kNItems];
        #pragma unroll
        for (int i = 0; i < kNItems; ++i) {
            dout_vals[i] = float(dout_vals_load[i]);
            delta_vals[i] = float(delta_vals_load[i]) + delta_bias;
            if constexpr (kDeltaSoftplus) {
                delta_vals[i] = delta_vals[i] <= 20.f ? log1pf(expf(delta_vals[i])) : delta_vals[i];
            }
        }
```

`delta__vals_load` will be correctly zeroed out, however after the `softplus` and `delta_bias` it will no longer be 0. This is bad because when we are calculating $\exp(\Delta A)$ in the following lines of code, we rely on the delta being zeroed out.

```CUDA
      const float delta_a_exp = exp2f(delta_vals[i] * A_scaled);
```

This then means that our propagation of gradients will be off. I have used `cuda-gdb` to check and ensure that this is happening. This doesn't seem to affect the gradients too much in `tests`, however, it could still be an issue that isn't covered by the test suite.

This can be fixed fairly simply by applying the following changes to the backward pass.

```CUDA
        for (int i = 0; i < kNItems; ++i) {
            dout_vals[i] = float(dout_vals_load[I]);
            if (threadIdx.x * kNItems + I < params.seqlen - chunk * kChunkSize) {
                  delta_vals[i] = float(delta_vals_load[i]) + delta_bias;
                  if constexpr (kDeltaSoftplus) {
                     delta_vals[i] = delta_vals[i] <= 20.f ? log1pf(expf(delta_vals[i])) : delta_vals[i];
                 }
            }
        }
```

**Comment by Hprairie** [comment]
I guess to clarify, we want delta to be 0 so that $\exp(\Delta A) = \exp(0) =1$. When delta bias and soft plus are used, then $\Delta \not= 0$ is almost guaranteed.

**Comment by Leopold2333** [comment]
> Hey, I think I found a bug in selective_scan_bwd_kernel.cuh . When we are propagating gradients backward we are moving in a reverse scan and thus it's important of exp⁡(ΔA) to be 1 for threads with items over the sequence length. This will be the case when `delta_softplus=False` and `delta_bias=None`, however, this will not be the case when either of these is enabled. We can see that this true by looking at the following lines of code.
> 
> ```cuda-c++
>         load_input<Ktraits>(delta, delta_vals_load, smem_load, params.seqlen - chunk * kChunkSize);
>         // Will reload delta at the same location if kDeltaSoftplus
>         if constexpr (!kDeltaSoftplus) { delta -= kChunkSize; }
> 
>         float dout_vals[kNItems], delta_vals[kNItems];
>         #pragma unroll
>         for (int i = 0; i < kNItems; ++i) {
>             dout_vals[i] = float(dout_vals_load[i]);
>             delta_vals[i] = float(delta_vals_load[i]) + delta_bias;
>             if constexpr (kDeltaSoftplus) {
>                 delta_vals[i] = delta_vals[i] <= 20.f ? log1pf(expf(delta_vals[i])) : delta_vals[i];
>             }
>         }
> ```
> 
> `delta__vals_load` will be correctly zeroed out, however after the `softplus` and `delta_bias` it will no longer be 0. This is bad because when we are calculating exp⁡(ΔA) in the following lines of code, we rely on the delta being zeroed out.
> 
> ```cuda-c++
>       const float delta_a_exp = exp2f(delta_vals[i] * A_scaled);
> ```
> 
> This then means that our propagation of gradients will be off. I have used `cuda-gdb` to check and ensure that this is happening. This doesn't seem to affect the gradients too much in `tests`, however, it could still be an issue that isn't covered by the test suite.
> 
> This can be fixed fairly simply by applying the following changes to the backward pass.
> 
> ```cuda-c++
>         for (int i = 0; i < kNItems; ++i) {
>             dout_vals[i] = float(dout_vals_load[I]);
>             if (threadIdx.x * kNItems + I < params.seqlen - chunk * kChunkSize) {
>                   delta_vals[i] = float(delta_vals_load[i]) + delta_bias;
>                   if constexpr (kDeltaSoftplus) {
>                      delta_vals[i] = delta_vals[i] <= 20.f ? log1pf(expf(delta_vals[i])) : delta_vals[i];
>                  }
>             }
>         }
> ```

Is this really ok? I change the corresponding part and reprogram the `selective_scan`, when I run this reprogrammed Mamba on my PC (with NVIDIA 40 series GPU), everything is OK, however, when I run the reprogrammed Mamba on some previous GPUs (I rented some NVIDIA 20 series GPUs), the loss gradients seems not been properly backwarded. After the `loss.backward()` process of the first minibatch, some layers will output 'nan' values. I'm confused, does the threads work differently for different GPUs?

**Comment by Hprairie** [comment]
I made a small error in the new code, which I have now fixed. Try with the new code.

**Comment by Leopold2333** [comment]
> I made a small error in the new code, which I have now fixed. Could you try with the latest code? The error could come from differences in how these GPUs fill empty vectors.

Oh, that's OK~
